src/Microsoft.VisualStudio.Composition/Configuration/Formatters/MessagePackSerializerContext.cs

```diff
@@ -0,0 +1,161 @@
+    using MessagePack;
+    using MessagePack.Formatters;
+    using MessagePack.Resolvers;
+    using Microsoft.VisualStudio.Composition.Formatter;
+
+    internal class MessagePackSerializerContext : MessagePackSerializerOptions, IDisposable
+    {
+        public MessagePackSerializerContext(IFormatterResolver resolver, Resolver compositionResolver)
+            : base(GetIFormatterResolver(resolver))
+        {
+            this.CompositionResolver = compositionResolver;
+        }
+
+        public Resolver CompositionResolver { get; }
+
+        public void Dispose()
+        {
+            GC.SuppressFinalize(this);
+        }
+
+        private static IFormatterResolver GetIFormatterResolver(IFormatterResolver resolver) =>
+            CompositeResolver.Create(
+                new IMessagePackFormatter[]
+                {
+                    new IgnoreFormatter<CancellationToken>(),
+                    new IgnoreFormatter<Task>(),
+                    ComposableCatalogFormatter.Instance,
+                    ComposablePartDefinitionFormatter.Instance,
+                    ExportDefinitionFormatter.Instance,
+                    FieldRefFormatter.Instance,
+                    ImportDefinitionBindingFormatter.Instance,
+                    ImportDefinitionFormatter.Instance,
+                    ImportMetadataViewConstraintFormatter.Instance,
+                    ExportMetadataValueImportConstraintFormatter.Instance,
+                    ExportTypeIdentityConstraintFormatter.Instance,
+                    MetadataDictionaryFormatter.Instance,
+                    MethodRefFormatter.Instance,
+                    ParameterRefFormatter.Instance,
+                    PartCreationPolicyConstraintFormatter.Instance,
+                    PropertyRefFormatter.Instance,
+                    RuntimeCompositionFormatter.Instance,
+                    RuntimeExportFormatter.Instance,
+                    RuntimeImportFormatter.Instance,
+                    RuntimePartFormatter.Instance,
+                    new StringInterningFormatter(),
+                },
+                new IFormatterResolver[]
+                {
+                     new DedupingResolver(resolver),
+                });
+
+        private class DedupingResolver : IFormatterResolver
+        {
+            private const sbyte ReferenceExtensionTypeCode = 1;
+            private readonly Dictionary<Type, IMessagePackFormatter> dedupingFormatters = new();
+            private readonly List<object?> deserializedObjects = new();
+            private readonly IFormatterResolver inner;
+            private readonly Dictionary<object, int> serializedObjects = new();
+            private int serializingObjectCounter;
+
+            internal DedupingResolver(IFormatterResolver inner)
+            {
+                this.inner = inner;
+            }
+
+            public IMessagePackFormatter<T>? GetFormatter<T>()
+            {
+                if (!typeof(T).IsValueType)
```

Comment:
I'm a touch worried that this is too aggressive a check. There are so many reference objects being serialized, and most of them are unlikely to ever appear twice. Yet it adds cost and size to every object in order to skip it later, which may never happen.
Also, it'll malfunction if `Equals` is poorly overridden. 
I wonder if instead we should test for specific types. Perhaps the same types that use `TryPrepareDeserializeReusableObject` in the `SerializationContextBase` class.